### ISSUE TITLE
issue: v7.22.5 brakes the way undefined form values are handled

### ISSUE THREAD

**Issue by codeaid** [react-hook-form/react-hook-form#7645]
### Version Number

>=7.22.5

### Codesandbox/Expo snack

https://stackblitz.com/edit/react-hook-form-bug?file=Form.tsx

### Steps to reproduce

I have created a component in the sandbox that replicates behaviour of the `Select` (dropdown) component that I have in my project. The idea is nothing new - you have a list of options in a dropdown, each of them has a value and you just pick one of them. Additionally, you can have an option at the top which indicates no selection (e.g. "All" or "None" depending on the context), whose value is `undefined` instead of it being a string.

As an example - it could be a list of colors to filter something by, e.g. "Red", "Blue", "Green" and one option above "All", which would indicate that no specific color should be used. That "All" option's value is `undefined`.

To demonstrate the issue I'm having:

1. Run the sandbox and ensure that the `react-hook-form` library version is set to `7.22.4` in the dependency panel on the left (important!). If it's not, paste `react-hook-form@7.22.4` into the package name field and press Enter.
2. Wait for the page to update and note that the `Form` component is being rendered with the following defaults in the `index.tsx` file (which are also reflected by the blue highlighted buttons in the panel on the right):
```jsx
<Form defaultValue1="one" defaultValue2={undefined} />
```
3. Next, observe how you can toggle both of the options between "one", "two" and "undefined" back and forth without any issues (you can also open the console to see some additional debug output). 
4. Now, change the version of the library to `7.22.5` by pasting `react-hook-form@7.22.5` into the package name field and pressing Enter.
5. Wait for the page to reload and try repeating the same option changes as before.
6. Notice how the top option that defaulted to `one` cannot be toggled to `undefined` anymore yet the second option, whose default value was `undefined` can.

I don't know what changed exactly between these two minor versions but this has broken all of my dropdown components as there is now no way of selecting the "All" options if the initial value of the form field was anything else but `undefined`.

I tried a few versions post 7.22.5 and they all exhibit the same issue.

### Expected behaviour

Form fields whose values can be 'undefined' should work the same way as they did prior to v7.22.5.

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
Just to add some additional info, notice how with the "new" 7.22.5 behavior, the value1 Controlled component doesn't rerender, which leads to it's value not being set to `undefined`. As seen in the Values object log, the `value1` form field itself does change to `undefined`.

https://user-images.githubusercontent.com/16290753/150844424-066f2beb-57a8-4c8f-96b1-21f1dd36bb70.mp4

I'd suggest to use `null` value as the default "none" / "all" option instead. It consistently works across all RHF versions and goes along with JS definitions of values

**Comment by codeaid** [comment]
@Moshyfawn Using `null` would've been my next attempt but that means refactoring quite a bit of logic in my code, including changing various types from `prop?: string` to `prop: string | null`, which I really don't want to do.

I was hoping this `undefined` behaviour is actually a bug so curious to see what the author says before I start refactoring it all.

**Comment by Moshyfawn** [comment]
Basically, [this check](https://github.com/react-hook-form/react-hook-form/blob/master/src/useWatch.ts#L88-L90) is what's breaking your use-case. If the new value is `undefined`, the field's `defaultValue` is used. You can see it fully if you select the option that isn't the default one _(option "two" in your case)_ and then try to select the "undefined" one. The value is set to "one" instead

**Comment by bluebill1049** [comment]
`undefined` need to be avoided, it is used to fall back to defaultValue or `defautlValues` since hook form does allow the user to provide inline defaultValue/defaultChecked.  The original fix was introduced to fix bug-related. See this PR for more detail: https://github.com/react-hook-form/react-hook-form/pull/7360

**Comment by wa-ri** [comment]
I have the same problem [CodeSandBox](https://codesandbox.io/s/usecontroller-forked-16xti?file=/src/App.js)

In `MyInput`, I can solve it through `onChange(null)`, but I cannot clear value in `MySelect`

**Comment by bluebill1049** [comment]
> I have the same problem [CodeSandBox](https://codesandbox.io/s/usecontroller-forked-16xti?file=/src/App.js)
> 
> In `MyInput`, I can solve it through `onChange(null)`, but I cannot clear value in `MySelect`

It's a text input, right? why not an empty string?

**Comment by wa-ri** [comment]
> > I have the same problem [CodeSandBox](https://codesandbox.io/s/usecontroller-forked-16xti?file=/src/App.js)
> > In `MyInput`, I can solve it through `onChange(null)`, but I cannot clear value in `MySelect`
> 
> It's a text input, right? why not an empty string?

yes, I can set an empty string or null value to clear the text input, but I cannot clear the select

**Comment by Moshyfawn** [comment]
> yes, I can set an empty string or null value to clear the text input, but I cannot clear the select
> record.mov

That's exactly what you're doing: clearing the field to it's default value. [The Antd Select component](https://ant.design/components/select/) passes `undefined` to `onChange` callback when the clear event happens
<img width="775" alt="image" src="https://user-images.githubusercontent.com/16290753/150954713-741e92e1-0f61-4d38-883e-c3f1597ccd74.png">

**Comment by wa-ri** [comment]
> > yes, I can set an empty string or null value to clear the text input, but I cannot clear the select
> > record.mov
> 
> That's exactly what you're doing: clearing the field to it's default value which is `Role: 1`. [The Antd Select component](https://ant.design/components/select/) passes `undefined` to `onChange` callback when the clear event happens <img alt="image" width="775" src="https://user-images.githubusercontent.com/16290753/150954713-741e92e1-0f61-4d38-883e-c3f1597ccd74.png">
> 
> You might wanna handle the `undefined` `onChange` like
> 
> ```js
> onChange={(value) =>
>   if (!value) {
>     field.onChange(null)
>   }
> }
> ```
> 
> Or something along these lines

In my scenario, null is an alternative value. Is there any other solution apart from changing null to 0, converting it to null and submiting it to the backend? I don't want null to be this complicated.
![image](https://user-images.githubusercontent.com/16372524/150967297-55b18612-8d5d-45dd-89a1-b11040e53011.png)
